### DISCUSSION TITLE
how do I avoid allowing the str default for more specific Literal overloads

### DISCUSSION THREAD

**Discussion by StarrFox** [microsoft/pyright#7229] [Q&A] [answered]
given the code
```py
from typing import Any, overload, Literal

@overload
def abc(format: Literal["?"], value: bool) -> None: ...

@overload
def abc(format: str, value: Any) -> None: ...

def abc(format: str, value: Any) -> None:
    pass

abc("?", True) # good
abc("def", "def") # good
abc("?", "def") # bad (should error)
```

how do I make the third usage an error but still allow usage 2? 

if format is ? then value must be bool but if format isn't ? then value can be Any

my use-case for this behavior is [this](https://github.com/StarrFox/memobj/blob/f09987eed0db4063565e5b517ddc9f9e11513918/memobj/process/base.py#L275-L349) where I want to type value based on struct format strings but still allow variables to be passed for format_string (non-literals)

**Comment by erictraut** [comment, marked as answer]
> how do I make the third usage an error but still allow usage 2?

There's no way to do this with the static type system. If you're able to change the interface, I recommend switching away from `"?"` as a sentinel value. It's problematic because `Literal["?"]` is a subtype of `str`. A sentinel value of `None` would be a better choice here.

```python
from typing import Any, overload, Literal

@overload
def abc(format: None, value: bool) -> None: ...

@overload
def abc(format: str, value: Any) -> None: ...

def abc(format: str | None, value: Any) -> None:
    pass

abc(None, True)  # good
abc("def", "def")  # good
abc(None, "def")  # bad
```

**Reply by StarrFox** [reply]
I see, basically the thing I'm typing is [struct.pack](https://docs.python.org/3/library/struct.html#struct.pack) but with only 1 value

I might try and use an enum instead of a string in order to solve this
